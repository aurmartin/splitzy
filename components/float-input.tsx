import React from "react";
import TextInput from "./text-input";
import type { TextInputProps } from "react-native";

interface FloatInputProps extends Omit<TextInputProps, "onChange" | "value"> {
  value: number;
  onChange: (value: number) => void;
}

const FloatInput = (props: FloatInputProps) => {
  const { value, onChange, ...rest } = props;
  const [text, setText] = React.useState(value.toString());

  React.useEffect(() => {
    setText((text) => {
      if (value !== parseFloat(text)) {
        return value.toString();
      }

      return text;
    });
  }, [value]);

  const handleChange = (textValue: string) => {
    if (!textValue.match(/^([0-9]{1,})?(\.)?([0-9]{1,})?$/)) {
      return;
    }

    let parsedValue = 0;
    if (textValue && parseFloat(textValue)) {
      parsedValue = parseFloat(textValue);
    }

    onChange(parsedValue);
    setText((prevText) => {
      if (prevText === "0") {
        return parsedValue.toString();
      }
      return textValue;
    });
  };

  return (
    <TextInput
      testID="float-input"
      value={text}
      onChangeText={handleChange}
      keyboardType="numeric"
      placeholder="0.00"
      {...rest}
    />
  );
};

export { FloatInput, type FloatInputProps };
