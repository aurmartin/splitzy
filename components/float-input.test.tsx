import * as React from "react";
import { render, screen, userEvent } from "@testing-library/react-native";
import { FloatInput } from "./float-input";

describe("FloatInput", () => {
  it("renders correctly", () => {
    const mockOnChange = jest.fn();
    const value = 10.0;

    render(<FloatInput value={value} onChange={mockOnChange} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("handles valid number inputs correctly", async () => {
    const user = userEvent.setup();

    const mockOnChange = jest.fn();
    const value = 10.0;

    render(<FloatInput value={value} onChange={mockOnChange} />);

    const input = screen.getByTestId("float-input");

    // Test whole numbers
    await user.clear(input);
    await user.type(input, "25");
    expect(mockOnChange).toHaveBeenCalledWith(25);

    // Test decimal numbers
    await user.clear(input);
    await user.type(input, "25.50");
    expect(mockOnChange).toHaveBeenCalledWith(25.5);

    // Test starting with decimal
    await user.clear(input);
    await user.type(input, ".5");
    expect(mockOnChange).toHaveBeenCalledWith(0.5);
  });

  it("ignores invalid number inputs", async () => {
    const user = userEvent.setup();

    const mockOnChange = jest.fn();
    const value = 10.0;

    render(<FloatInput value={value} onChange={mockOnChange} />);

    const input = screen.getByTestId("float-input");

    // Test invalid characters
    await user.type(input, "abc");
    expect(input.props.value).toBe("10");
    expect(mockOnChange).not.toHaveBeenCalled();
  });

  it("updates display value when prop changes", async () => {
    const mockOnChange = jest.fn();
    render(<FloatInput value={10.0} onChange={mockOnChange} />);

    const input = screen.getByTestId("float-input");

    expect(input.props.value).toBe("10");

    screen.rerender(<FloatInput value={20.0} onChange={mockOnChange} />);
    expect(input.props.value).toBe("20");
  });
});
