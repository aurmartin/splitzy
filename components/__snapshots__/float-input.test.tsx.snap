// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatInput renders correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 8,
        "flexDirection": "row",
        "height": 55,
        "paddingHorizontal": 16,
      },
      undefined,
    ]
  }
>
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="0.00"
      placeholderTextColor="#888"
      style={
        {
          "flex": 1,
          "fontSize": 16,
          "paddingBottom": 0,
          "paddingTop": 0,
          "textAlign": "left",
        }
      }
      testID="float-input"
      value="10"
    />
  </View>
</View>
`;
